impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtBuilder.java

```diff
@@ -693,14 +693,13 @@ private String encrypt(final Payload content, final Key key, final Provider keyP
         Assert.stateNotNull(keyAlgFunction, "KeyAlgorithm function cannot be null.");
         assertPayloadEncoding("JWE");
 
-        InputStream plaintext;
+        ByteArrayOutputStream out = new ByteArrayOutputStream(4096);
         if (content.isClaims()) {
-            ByteArrayOutputStream out = new ByteArrayOutputStream(4096);
             writeAndClose("JWE Claims", content, out);
-            plaintext = Streams.of(out.toByteArray());
         } else {
-            plaintext = content.toInputStream();
+            writeAndClose("JWE Content", content, out);
```

Comment:
We have to be careful here: we don't want to write the plaintext Inputstream written to an intermediate output stream (and then converted back into an input stream) requiring a roundtrip that isn't necessary if compression is not enabled.

This logic should look more like that found in the `sign` method:

https://github.com/jwtk/jjwt/blob/c673b76ef5b74ddaf51b4e80fc632ddb62b7e29a/impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtBuilder.java#L598-L605

If possible, it'd be ideal if we can move this logic to a helper method that is shared by both `sign` and `encrypt`

Reply:
Makes sense.

I wonder though if we should also be a bit more clever with the initial byte array size for ByteArrayOutputStream. Seems a bit wasteful to do 8 KB byte array allocation if the content is only few hundred bytes. At least if the payload has byte array / string, we could use the initial payload size as the starting point, as even with compression, we shouldn't (at least usually) go above that size.

Reply:
Done